x-pack/plugins/stack_alerts/server/alert_types/geo_containment/geo_containment.ts

```diff
@@ -106,32 +91,54 @@ export function getActiveEntriesAndGenerateAlerts(
   shapesIdsNamesMap: Record<string, unknown>,
   currIntervalEndTime: Date
 ) {
-  const allActiveEntriesMap: Map<string, LatestEntityLocation> = new Map([
-    ...Object.entries(prevLocationMap || {}),
+  const allActiveEntriesMap: Map<string, LatestEntityLocation[]> = new Map([
+    ...prevLocationMap,
     ...currLocationMap,
   ]);
-  allActiveEntriesMap.forEach(({ location, shapeLocationId, dateInShape, docId }, entityName) => {
-    const containingBoundaryName = shapesIdsNamesMap[shapeLocationId] || shapeLocationId;
-    const context = {
-      entityId: entityName,
-      entityDateTime: dateInShape ? new Date(dateInShape).toISOString() : null,
-      entityDocumentId: docId,
-      detectionDateTime: new Date(currIntervalEndTime).toISOString(),
-      entityLocation: `POINT (${location[0]} ${location[1]})`,
-      containingBoundaryId: shapeLocationId,
-      containingBoundaryName,
-    };
-    const alertInstanceId = `${entityName}-${containingBoundaryName}`;
-    if (shapeLocationId === OTHER_CATEGORY) {
+  allActiveEntriesMap.forEach((locationsArr, entityName) => {
+    // Generate alerts
+    locationsArr.forEach(({ location, shapeLocationId, dateInShape, docId }) => {
+      const context = {
+        entityId: entityName,
+        entityDateTime: dateInShape ? new Date(dateInShape).toISOString() : null,
+        entityDocumentId: docId,
+        detectionDateTime: new Date(currIntervalEndTime).toISOString(),
+        entityLocation: `POINT (${location[0]} ${location[1]})`,
+        containingBoundaryId: shapeLocationId,
+        containingBoundaryName: shapesIdsNamesMap[shapeLocationId] || shapeLocationId,
+      };
+      const alertInstanceId = `${entityName}-${context.containingBoundaryName}`;
+      if (shapeLocationId !== OTHER_CATEGORY) {
+        alertInstanceFactory(alertInstanceId).scheduleActions(ActionGroupId, context);
+      }
+    });
+    let otherIndex;
+    if (locationsArr[0].shapeLocationId === OTHER_CATEGORY) {
       allActiveEntriesMap.delete(entityName);
+    } else if (
+      locationsArr.some(({ shapeLocationId }, index) => {
+        otherIndex = index;
+        return shapeLocationId === OTHER_CATEGORY;
+      })
+    ) {
```

Comment:
this block seems wonky? . Where does this sit exactly in the flow?

```
if(){
}else if () {
} {
//??
} elsee {
}
```

Reply:
I believe it's correct, it's just a bulky conditional statement that uses the `array.some` method and also grabs the index while we're in there to save time in the block below it. Then again I used to write Java for a living, so maybe my tolerance for doing something in 20 lines that should be 2 is too high. If you re-read it and it still looks bad, maybe I'll see if I can restructure that whole if/else block. Just let me know!

Reply:
Yeah, it's a little dense :), and produces some side-effects in that conditional. 

Maybe something along:

```
    if (locationsArr[0].shapeLocationId === OTHER_CATEGORY) {
      allActiveEntriesMap.delete(entityName);
     return;
    } 

    const otherIndex = locationsArr.findIndex({ shapeLocationId }, index) => { return shapeLocationId === OTHER_CATEGORY;});
    if (otherIndex >= 0) {
      const afterOtherLocationsArr = locationsArr.slice(0, otherIndex);
      allActiveEntriesMap.set(entityName, afterOtherLocationsArr);
    } else {
      allActiveEntriesMap.set(entityName, locationsArr);
    }

```